metals/src/main/scala/scala/meta/internal/metals/ScalaTarget.scala

```diff
@@ -29,28 +29,28 @@ case class ScalaTarget(
 
   def isSbt = sbtVersion.isDefined
 
-  def dialect(path: AbsolutePath): Dialect = {
-    scalaVersion match {
-      case _ if info.isSbtBuild && path.isSbt => Sbt
-      case other =>
-        val dialect =
-          ScalaVersions.dialectForScalaVersion(other, includeSource3 = false)
-        dialect match {
-          case Scala213 if containsSource3 =>
-            Scala213Source3
-          case Scala212 if containsSource3 =>
-            Scala212Source3
-          case Scala3
-              if scalac
-                .getOptions()
-                .asScala
-                .exists(_.startsWith("-Ykind-projector")) =>
-            dialect.withAllowStarAsTypePlaceholder(true)
-          case Scala3 =>
-            // set to false since this needs an additional compiler option
-            Scala3.withAllowStarAsTypePlaceholder(false)
-          case other => other
-        }
+  def dialect(path: AbsolutePath): Dialect =
+    if (info.isSbtBuild && path.isSbt) Sbt
+    else scalaDialect
+
+  private def scalaDialect: Dialect = {
+    def kindProjector = options.exists(_.matches("-[XY]kind-projector.*"))
+    def scalaFuture = options.exists(
+      _.matches("-language:experimental.(captureChecking|modularity|into)")
```

Comment:
```suggestion
      _.startsWith("-language:experimental.")
```
would not be enough? I think also it's a bit more efficient than regexes. Otherwise I would compile the regex beforehand